autorelay: racy TestBackoff

```
=== RUN   TestBackoff
  ==================
  WARNING: DATA RACE
  Read at 0x00c0017393c8 by goroutine 251:
    github.com/libp2p/go-libp2p/p2p/host/autorelay_test.TestBackoff()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/autorelay_test.go:267 +0xa54
    testing.tRunner()
        /opt/hostedtoolcache/go/1.18.5/x64/src/testing/testing.go:1439 +0x213
    testing.(*T).Run.func1()
        /opt/hostedtoolcache/go/1.18.5/x64/src/testing/testing.go:1486 +0x47
  
  Previous write at 0x00c0017393c8 by goroutine 261:
    github.com/libp2p/go-libp2p/p2p/host/autorelay_test.TestBackoff.func3()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/autorelay_test.go:246 +0x58
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).findNodes()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:238 +0x7e4
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).background.func2()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:112 +0x9e
  
  Goroutine 251 (running) created at:
    testing.(*T).Run()
        /opt/hostedtoolcache/go/1.18.5/x64/src/testing/testing.go:1486 +0x724
    testing.runTests.func1()
        /opt/hostedtoolcache/go/1.18.5/x64/src/testing/testing.go:1839 +0x99
    testing.tRunner()
        /opt/hostedtoolcache/go/1.18.5/x64/src/testing/testing.go:1[439](https://github.com/libp2p/go-libp2p/runs/8150958283?check_suite_focus=true#step:9:441) +0x213
    testing.runTests()
        /opt/hostedtoolcache/go/1.18.5/x64/src/testing/testing.go:1837 +0x7e4
    testing.(*M).Run()
        /opt/hostedtoolcache/go/1.18.5/x64/src/testing/testing.go:1719 +0xa71
    main.main()
        _testmain.go:67 +0x2e4
  
  Goroutine 261 (running) created at:
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).background()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:110 +0x25e
    github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).Start.func1()
        /home/runner/work/go-libp2p/go-libp2p/p2p/host/autorelay/relay_finder.go:658 +0x9e
  ==================
      testing.go:1312: race detected during execution of test
  --- FAIL: TestBackoff (0.14s)
  === RUN   TestStaticRelays
  --- PASS: TestStaticRelays (0.47s)
  === RUN   TestRelayV1
  === RUN   TestRelayV1/relay_v1_support_disabled
  === RUN   TestRelayV1/relay_v1_support_enabled
  --- PASS: TestRelayV1 (0.42s)
      --- PASS: TestRelayV1/relay_v1_support_disabled (0.29s)
      --- PASS: TestRelayV1/relay_v1_support_enabled (0.13s)
  === RUN   TestConnectOnDisconnect
  --- PASS: TestConnectOnDisconnect (0.51s)
  === RUN   TestMaxAge
  --- PASS: TestMaxAge (1.40s)
  === CONT  
      testing.go:1312: race detected during execution of test
  FAIL
  FAIL	github.com/libp2p/go-libp2p/p2p/host/autorelay	6.362s

--- a/p2p/host/autorelay/autorelay_test.go
+++ b/p2p/host/autorelay/autorelay_test.go
@@ -239,11 +239,11 @@ func TestBackoff(t *testing.T) {
 		str.Reset()
 	})
 
-	var counter int
+	var counter int32 // to be used atomically
 	h := newPrivateNode(t,
 		autorelay.WithPeerSource(func(int) <-chan peer.AddrInfo {
 			// always return the same node, and make sure we don't try to connect to it too frequently
-			counter++
+			atomic.AddInt32(&counter, 1)
 			peerChan := make(chan peer.AddrInfo, 1)
 			peerChan <- peer.AddrInfo{ID: r.ID(), Addrs: r.Addrs()}
 			close(peerChan)
@@ -264,7 +264,7 @@ func TestBackoff(t *testing.T) {
 	}
 	cl.Add(backoff / 2)
 	require.Eventually(t, func() bool { return atomic.LoadInt32(&reservations) == 2 }, 3*time.Second, 20*time.Millisecond)
-	require.Less(t, counter, 100) // just make sure we're not busy-looping
+	require.Less(t, int(atomic.LoadInt32(&counter)), 100) // just make sure we're not busy-looping
 	require.Equal(t, 2, int(atomic.LoadInt32(&reservations)))
 }
 

